File: packages/nxuskit-go/pipeline/loader.go
// Package pipeline provides loading from JSON and YAML files.
package pipeline

import (
	"encoding/json"
	"errors"
	"fmt"
	"log/slog"
	"os"
	"path/filepath"
	"strings"

	"gopkg.in/yaml.v3"
)

// LoadError represents errors that can occur during pipeline loading.
type LoadError struct {
	Code    string
	Message string
	Cause   error
}

func (e *LoadError) Error() string {
	if e.Cause != nil {
		return fmt.Sprintf("%s: %v", e.Message, e.Cause)
	}
	return e.Message
}

func (e *LoadError) Unwrap() error {
	return e.Cause
}

var (
	// ErrUnsupportedFormat indicates an unsupported file format.
	ErrUnsupportedFormat = errors.New("unsupported file format")
	// ErrLicenseRequired indicates Pro license is required for CLIPS features.
	ErrLicenseRequired = errors.New("pro license required for CLIPS features")
)

// fileFormat represents the detected file format.
type fileFormat int

const (
	formatJSON fileFormat = iota
	formatYAML
)

// detectFormat determines the file format from the extension.
func detectFormat(path string) (fileFormat, error) {
	ext := strings.ToLower(filepath.Ext(path))
	switch ext {
	case ".json":
		return formatJSON, nil
	case ".yaml", ".yml":
		return formatYAML, nil
	default:
		return 0, &LoadError{
			Code:    "UNSUPPORTED_FORMAT",
			Message: fmt.Sprintf("unsupported file format: %s. Expected .json or .yaml/.yml", ext),
		}
	}
}

// LoadPipeline loads a pipeline definition from a file.
// Automatically detects JSON or YAML format based on file extension.
func LoadPipeline(path string) (*PipelineDefinition, error) {
	slog.Info("Loading pipeline definition", "path", path)

	format, err := detectFormat(path)
	if err != nil {
		return nil, err
	}

	data, err := os.ReadFile(path)
	if err != nil {
		return nil, &LoadError{
			Code:    "IO_ERROR",
			Message: "failed to read file",
			Cause:   err,
		}
	}

	pipeline, err := LoadPipelineFromBytes(data, format)
	if err != nil {
		return nil, err
	}

	slog.Info("Pipeline loaded successfully",
		"pipeline_id", pipeline.ID,
		"pipeline_name", pipeline.Name,
		"stage_count", len(pipeline.Stages),
	)

	return pipeline, nil
}

// LoadPipelineFromBytes loads a pipeline definition from byte content.
func LoadPipelineFromBytes(data []byte, format fileFormat) (*PipelineDefinition, error) {
	var pipeline PipelineDefinition

	switch format {
	case formatJSON:
		if err := json.Unmarshal(data, &pipeline); err != nil {
			return nil, &LoadError{
				Code:    "JSON_ERROR",
				Message: "failed to parse JSON",
				Cause:   err,
			}
		}
	case formatYAML:
		if err := yaml.Unmarshal(data, &pipeline); err != nil {
			return nil, &LoadError{
				Code:    "YAML_ERROR",
				Message: "failed to parse YAML",
				Cause:   err,
			}
		}
	}

	// Check for CLIPS stages that require Pro license
	if err := checkClipsLicense(&pipeline); err != nil {
		return nil, err
	}

	return &pipeline, nil
}

// checkClipsLicense checks if CLIPS stages are present and license is available.
// In the free tier, this returns an error for CLIPS stages.
func checkClipsLicense(pipeline *PipelineDefinition) error {
	// This is a stub - actual license checking would be implemented
	// based on the licensing system
	for _, stage := range pipeline.Stages {
		switch stage.Type {
		case StageTypeClipsEval, StageTypeClipsGen, StageTypeHybrid:
			// In production, check for actual Pro license
			// For now, we allow all stages (license check returns true)
			// Uncomment below to enforce license:
			// slog.Warn("CLIPS stage detected but Pro license not verified",
			// 	"stage_id", stage.ID,
			// 	"stage_type", stage.Type,
			// )
			// return ErrLicenseRequired
		}
	}
	return nil
}

// LoadPipelineFromJSON loads a pipeline definition from a JSON string.
func LoadPipelineFromJSON(content string) (*PipelineDefinition, error) {
	return LoadPipelineFromBytes([]byte(content), formatJSON)
}

// LoadPipelineFromYAML loads a pipeline definition from a YAML string.
func LoadPipelineFromYAML(content string) (*PipelineDefinition, error) {
	return LoadPipelineFromBytes([]byte(content), formatYAML)
}

// SavePipeline saves a pipeline definition to a file.
// Automatically detects JSON or YAML format based on file extension.
func SavePipeline(pipeline *PipelineDefinition, path string) error {
	slog.Info("Saving pipeline definition", "path", path)

	format, err := detectFormat(path)
	if err != nil {
		return err
	}

	var data []byte
	switch format {
	case formatJSON:
		data, err = json.MarshalIndent(pipeline, "", "  ")
	case formatYAML:
		data, err = yaml.Marshal(pipeline)
	}

	if err != nil {
		return &LoadError{
			Code:    "SERIALIZE_ERROR",
			Message: "failed to serialize pipeline",
			Cause:   err,
		}
	}

	if err := os.WriteFile(path, data, 0644); err != nil {
		return &LoadError{
			Code:    "IO_ERROR",
			Message: "failed to write file",
			Cause:   err,
		}
	}

	slog.Info("Pipeline saved successfully", "path", path)
	return nil
}

// PipelineToJSON serializes a pipeline definition to JSON.
func PipelineToJSON(pipeline *PipelineDefinition) (string, error) {
	data, err := json.MarshalIndent(pipeline, "", "  ")
	if err != nil {
		return "", err
	}
	return string(data), nil
}

// PipelineToYAML serializes a pipeline definition to YAML.
func PipelineToYAML(pipeline *PipelineDefinition) (string, error) {
	data, err := yaml.Marshal(pipeline)
	if err != nil {
		return "", err
	}
	return string(data), nil
}

File: packages/nxuskit-engine/crates/nxuskit-engine/src/pipeline/loader.rs
//! Pipeline loading from JSON and YAML files.
//!
//! This module provides functions to load pipeline definitions from
//! JSON or YAML files with automatic format detection.

use std::fs;
use std::path::Path;

use super::PipelineDefinition;

/// Errors that can occur during pipeline loading.
#[derive(Debug, thiserror::Error)]
pub enum LoadError {
    /// File I/O error.
    #[error("Failed to read file: {0}")]
    IoError(#[from] std::io::Error),

    /// JSON parsing error.
    #[error("Failed to parse JSON: {0}")]
    JsonError(#[from] serde_json::Error),

    /// YAML parsing error.
    #[error("Failed to parse YAML: {0}")]
    YamlError(#[from] serde_yaml::Error),

    /// Unsupported file format.
    #[error("Unsupported file format: {extension}. Expected .json or .yaml/.yml")]
    UnsupportedFormat { extension: String },

    /// License required for CLIPS features.
    #[error("Pro license required for CLIPS features")]
    LicenseRequired,
}

/// Result type for load operations.
pub type LoadResult<T> = Result<T, LoadError>;

/// Detect file format from extension.
fn detect_format(path: &Path) -> LoadResult<FileFormat> {
    let extension = path
        .extension()
        .and_then(|e| e.to_str())
        .unwrap_or("")
        .to_lowercase();

    match extension.as_str() {
        "json" => Ok(FileFormat::Json),
        "yaml" | "yml" => Ok(FileFormat::Yaml),
        _ => Err(LoadError::UnsupportedFormat { extension }),
    }
}

#[derive(Debug, Clone, Copy)]
enum FileFormat {
    Json,
    Yaml,
}

/// Load a pipeline definition from a file.
///
/// Automatically detects JSON or YAML format based on file extension.
///
/// # Arguments
///
/// * `path` - Path to the pipeline definition file (.json, .yaml, or .yml)
///
/// # Example
///
/// ```no_run
/// use nxuskit_engine::pipeline::load_pipeline;
///
/// let pipeline = load_pipeline("my-pipeline.json").unwrap();
/// println!("Loaded pipeline: {}", pipeline.name);
/// ```
pub fn load_pipeline<P: AsRef<Path>>(path: P) -> LoadResult<PipelineDefinition> {
    let path = path.as_ref();
    // Note: Consider adding tracing when available in workspace dependencies

    let format = detect_format(path)?;
    let content = fs::read_to_string(path)?;

    let pipeline = load_pipeline_from_str(&content, format)?;

    Ok(pipeline)
}

/// Load a pipeline definition from a string.
///
/// # Arguments
///
/// * `content` - The pipeline definition content
/// * `format` - The format of the content (JSON or YAML)
fn load_pipeline_from_str(content: &str, format: FileFormat) -> LoadResult<PipelineDefinition> {
    let pipeline: PipelineDefinition = match format {
        FileFormat::Json => serde_json::from_str(content)?,
        FileFormat::Yaml => serde_yaml::from_str(content)?,
    };

    Ok(pipeline)
}

/// Load a pipeline definition from a JSON string.
///
/// # Example
///
/// ```
/// use nxuskit_engine::pipeline::load_pipeline_from_json;
///
/// let json = r#"{
///     "id": "test",
///     "name": "Test Pipeline",
///     "stages": [],
///     "created_at": "2026-01-31T00:00:00Z",
///     "updated_at": "2026-01-31T00:00:00Z"
/// }"#;
///
/// let pipeline = load_pipeline_from_json(json).unwrap();
/// assert_eq!(pipeline.name, "Test Pipeline");
/// ```
pub fn load_pipeline_from_json(content: &str) -> LoadResult<PipelineDefinition> {
    load_pipeline_from_str(content, FileFormat::Json)
}

/// Load a pipeline definition from a YAML string.
///
/// # Example
///
/// ```
/// use nxuskit_engine::pipeline::load_pipeline_from_yaml;
///
/// let yaml = r#"
/// id: test
/// name: Test Pipeline
/// stages: []
/// created_at: "2026-01-31T00:00:00Z"
/// updated_at: "2026-01-31T00:00:00Z"
/// "#;
///
/// let pipeline = load_pipeline_from_yaml(yaml).unwrap();
/// assert_eq!(pipeline.name, "Test Pipeline");
/// ```
pub fn load_pipeline_from_yaml(content: &str) -> LoadResult<PipelineDefinition> {
    load_pipeline_from_str(content, FileFormat::Yaml)
}

/// Save a pipeline definition to a file.
///
/// Automatically detects JSON or YAML format based on file extension.
///
/// # Arguments
///
/// * `pipeline` - The pipeline definition to save
/// * `path` - Path to save the file (.json, .yaml, or .yml)
pub fn save_pipeline<P: AsRef<Path>>(pipeline: &PipelineDefinition, path: P) -> LoadResult<()> {
    let path = path.as_ref();

    let format = detect_format(path)?;

    let content = match format {
        FileFormat::Json => serde_json::to_string_pretty(pipeline)?,
        FileFormat::Yaml => serde_yaml::to_string(pipeline)?,
    };

    fs::write(path, content)?;

    Ok(())
}

/// Serialize a pipeline definition to JSON.
pub fn pipeline_to_json(pipeline: &PipelineDefinition) -> Result<String, serde_json::Error> {
    serde_json::to_string_pretty(pipeline)
}

/// Serialize a pipeline definition to YAML.
pub fn pipeline_to_yaml(pipeline: &PipelineDefinition) -> Result<String, serde_yaml::Error> {
    serde_yaml::to_string(pipeline)
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_json_round_trip() {
        let json = r#"{
            "id": "test-id",
            "name": "Test Pipeline",
            "stages": [
                {
                    "id": "stage1",
                    "name": "Stage 1",
                    "type": "llm",
                    "llm_config": {
                        "provider": "claude",
                        "model": "claude-3-sonnet",
                        "user_prompt": "Test"
                    }
                }
            ],
            "created_at": "2026-01-31T00:00:00Z",
            "updated_at": "2026-01-31T00:00:00Z"
        }"#;

        let pipeline = load_pipeline_from_json(json).unwrap();
        assert_eq!(pipeline.id, "test-id");
        assert_eq!(pipeline.name, "Test Pipeline");
        assert_eq!(pipeline.stages.len(), 1);

        // Round-trip
        let json_out = pipeline_to_json(&pipeline).unwrap();
        let pipeline2 = load_pipeline_from_json(&json_out).unwrap();
        assert_eq!(pipeline.id, pipeline2.id);
        assert_eq!(pipeline.name, pipeline2.name);
    }

    #[test]
    fn test_yaml_round_trip() {
        let yaml = r#"
id: test-id
name: Test Pipeline
stages:
  - id: stage1
    name: Stage 1
    type: llm
    llm_config:
      provider: claude
      model: claude-3-sonnet
      user_prompt: Test
created_at: "2026-01-31T00:00:00Z"
updated_at: "2026-01-31T00:00:00Z"
"#;

        let pipeline = load_pipeline_from_yaml(yaml).unwrap();
        assert_eq!(pipeline.id, "test-id");
        assert_eq!(pipeline.name, "Test Pipeline");
        assert_eq!(pipeline.stages.len(), 1);

        // Round-trip
        let yaml_out = pipeline_to_yaml(&pipeline).unwrap();
        let pipeline2 = load_pipeline_from_yaml(&yaml_out).unwrap();
        assert_eq!(pipeline.id, pipeline2.id);
        assert_eq!(pipeline.name, pipeline2.name);
    }

    #[test]
    fn test_format_detection() {
        assert!(matches!(
            detect_format(Path::new("test.json")),
            Ok(FileFormat::Json)
        ));
        assert!(matches!(
            detect_format(Path::new("test.yaml")),
            Ok(FileFormat::Yaml)
        ));
        assert!(matches!(
            detect_format(Path::new("test.yml")),
            Ok(FileFormat::Yaml)
        ));
        assert!(matches!(
            detect_format(Path::new("test.txt")),
            Err(LoadError::UnsupportedFormat { .. })
        ));
    }
}
